### PR TITLE
[WIP] webdataset prototype - class to preprocess image folder

### PR DESCRIPTION
**Pull request by glaringlee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50253 [WIP] webdataset prototype - class to preprocess image folder**
* #49353 webdataset prototype - GroupByFilenameIterableDataset
* #49140 webdataset prototype - RoutedDecoderIterableDataset
* #49026 webdataset prototype - ReadFilesFromZipIterableDataset
* #49000 webdataset prototype - ReadFilesFromTarIterableDataset

This is for now a super simple class to preprocess image folder.
The only functionality it currently support is to pair each image file with a label file (with category_id set to the name of the folder(s)) and convert everything into one tar file.

Differential Revision: [D26012931](https://our.internmc.facebook.com/intern/diff/D26012931)

### CODE REVIEW

**Review by ejguan — COMMENTED** [review]
I am not sure if you just want to use this class to make one tar of one folder for same category. Can we do it on parent folder? Just make one tar file per one sub-folder of one category. Then merge these tar files into one tar on parent folder.

File: torch/utils/data/datasets/imagefolder.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                bio.write(str.encode('{{"category_id": "{}"}}'.format(category_id)))
+                path_info = Path(pathname)
+                tinfo = tarfile.TarInfo(basename + ".json")
+                tinfo.size = bio.tell()
+                bio.seek(0)
+                tinfo.mtime = time.time()
+                tinfo.uname = path_info.owner()
+                tinfo.gname = path_info.group()
+                tinfo.mode = path_info.stat().st_mode & 0o0777
+                tarstream.addfile(tinfo, bio)
```
Comment (ejguan) [inline torch/utils/data/datasets/imagefolder.py:62]: If so, you can move this part out of for loop since all image files will have same category.

**Comment by glaringlee** [comment]
> I am not sure if you just want to use this class to make one tar of one folder for same category. Can we do it on parent folder? Just make one tar file per one sub-folder of one category. Then merge these tar files into one tar on parent folder.

Yes, understood. Currently, the entire of this stack does not support recursive operation which means we don't support hierarchy structured folder. The earlier versions in this stack do support, but then we decide to remove this support first to make everything simple, and we can add it later.

**Comment by ejguan** [comment]
> > I am not sure if you just want to use this class to make one tar of one folder for same category. Can we do it on parent folder? Just make one tar file per one sub-folder of one category. Then merge these tar files into one tar on parent folder.
> 
> @ejguan Yes, understood. Currently, the entire of this stack does not support recursive operation which means we don't support hierarchy structured folder. The earlier versions in this stack do support, but then we decide to remove this support first to make everything simple, and we can add it later.

Wait. Is this the one for testing?
BTW we can still do recursive here by just walking through different folders but save them into the root path of tar file. So we can actually have different label in the tar, therefore we can train model.

**Comment by glaringlee** [comment]
> > > I am not sure if you just want to use this class to make one tar of one folder for same category. Can we do it on parent folder? Just make one tar file per one sub-folder of one category. Then merge these tar files into one tar on parent folder.
> > 
> > 
> > @ejguan Yes, understood. Currently, the entire of this stack does not support recursive operation which means we don't support hierarchy structured folder. The earlier versions in this stack do support, but then we decide to remove this support first to make everything simple, and we can add it later.
> 
> Wait. Is this the one for testing?
> BTW we can still do recursive here by just walking through different folders but save them into the root path of tar file. So we can actually have different label in the tar, therefore we can train model.

@ejguan Make sense, I changed the code to accept a list of path and pack everything into one tar file for now. There is still problems, I think we ultimately need to enable recursively scan folders, since some of the folders are structured like this.
folder_with_folder_label
                     |--> images
                                |--> image files...
                     |--> readme.txt 
 Let's leave the code like this for now, and discuss this next week.

**Comment by glaringlee** [comment]
merged to torchdata already

**Pull request closed without merging**